examples/chip-tool/commands/clusters/DataModelLogger.h

```diff
@@ -34,7 +52,122 @@ class DataModelLogger
+        value[kEndpointIdKey]  = path.mEndpointId;
+        value[kAttributeIdKey] = path.mAttributeId;
+
+        return LogError(value, status);
+    }
+
+    static CHIP_ERROR LogCommandAsJSON(const chip::app::ConcreteCommandPath & path, chip::TLV::TLVReader * data)
+    {
+        VerifyOrReturnError(mJSONDelegate != nullptr, CHIP_NO_ERROR);
+
+        Json::Value value;
+        value[kClusterIdKey]  = path.mClusterId;
+        value[kEndpointIdKey] = path.mEndpointId;
+        value[kCommandIdKey]  = path.mCommandId;
+
+        chip::TLV::TLVReader reader;
+        reader.Init(*data);
+        ReturnErrorOnFailure(chip::TlvToJson(reader, value));
+
+        auto valueStr = chip::JsonToString(value);
+        return mJSONDelegate->LogJSON(valueStr.c_str());
+    }
+
+    static CHIP_ERROR LogErrorAsJSON(const chip::app::ConcreteCommandPath & path, const chip::app::StatusIB & status)
+    {
+        VerifyOrReturnError(mJSONDelegate != nullptr, CHIP_NO_ERROR);
+
+        Json::Value value;
+        value[kClusterIdKey]  = path.mClusterId;
+        value[kEndpointIdKey] = path.mEndpointId;
+        value[kCommandIdKey]  = path.mCommandId;
+
+        return LogError(value, status);
+    }
+
+    static CHIP_ERROR LogEventAsJSON(const chip::app::EventHeader & header, chip::TLV::TLVReader * data)
+    {
+        VerifyOrReturnError(mJSONDelegate != nullptr, CHIP_NO_ERROR);
+
+        Json::Value value;
+        value[kClusterIdKey]  = header.mPath.mClusterId;
+        value[kEndpointIdKey] = header.mPath.mEndpointId;
+        value[kEventIdKey]    = header.mPath.mEventId;
+
+        chip::TLV::TLVReader reader;
+        reader.Init(*data);
+        ReturnErrorOnFailure(chip::TlvToJson(reader, value));
+
+        auto valueStr = chip::JsonToString(value);
+        return mJSONDelegate->LogJSON(valueStr.c_str());
+    }
+
+    static CHIP_ERROR LogErrorAsJSON(const chip::app::EventHeader & header, const chip::app::StatusIB & status)
+    {
+        VerifyOrReturnError(mJSONDelegate != nullptr, CHIP_NO_ERROR);
+
+        Json::Value value;
+        value[kClusterIdKey]  = header.mPath.mClusterId;
+        value[kEndpointIdKey] = header.mPath.mEndpointId;
+        value[kEventIdKey]    = header.mPath.mEventId;
+
+        return LogError(value, status);
+    }
+
+    static CHIP_ERROR LogErrorAsJSON(const CHIP_ERROR & error)
+    {
+        Json::Value value;
+        chip::app::StatusIB status;
+        status.InitFromChipError(error);
+        return LogError(value, status);
+    }
+
+    static void SetJSONDelegate(DataModelLoggerJSONDelegate * delegate) { mJSONDelegate = delegate; }
+
 private:
+    static CHIP_ERROR LogError(Json::Value & value, const chip::app::StatusIB & status)
+    {
+        if (status.mClusterStatus.HasValue())
+        {
+            auto statusValue          = status.mClusterStatus.Value();
+            value[kErrorIdKey]        = kGenericFailure;
+            value[kClusterErrorIdKey] = statusValue;
```

Comment:
What if this is a cluster-specific success status?  Is the idea that we know this is only called for failure status?  If so, we should error out if that's not the case.

Reply:
My understanding is that cluster status error code results into a "Status::Failure" and the current expectation from the test suite is that a cluster status error code results into a "FAILURE". I believe we can revisit that later if that is not exactly right.
FTR, this code is only called in case where https://github.com/project-chip/connectedhomeip/blob/9fbcf1cef8c4a42761a7fc23588ca71d75904ff5/src/app/MessageDef/StatusIB.cpp#L137 does not returns `CHIP_NO_ERROR`.